lib/statsd/instrument/aggregator.rb

```diff
@@ -0,0 +1,246 @@
+        @flush_thread = Thread.new do
+          Thread.current.abort_on_exception = true
+          loop do
+            sleep(@flush_interval)
+            thread_healthcheck
+            flush
+          end
+        end
+
+        ObjectSpace.define_finalizer(
+          self,
+          self.class.finalize(@aggregation_state, @sink, @datagram_builders, @datagram_builder_class),
+        )
+      end
+
+      # Increment a counter by a given value and save it for later flushing.
+      # @param name [String] The name of the counter.
+      # @param value [Integer] The value to increment the counter by.
+      # @param tags [Hash{String, Symbol => String},Array<String>] The tags to attach to the counter.
+      # @param no_prefix [Boolean] If true, the metric will not be prefixed.
+      # @return [void]
+      def increment(name, value = 1, tags: [], no_prefix: false)
+        unless thread_healthcheck
+          sink << datagram_builder(no_prefix: no_prefix).c(name, value, CONST_SAMPLE_RATE, tags)
+          return
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, COUNT)
+
+        mutex.synchronize do
+          unless aggregation_state.key?(key)
+            aggregation_state[key] = {
+              type: COUNT,
+              name: name,
+              value: 0,
+              tags: tags,
+              no_prefix: no_prefix,
+            }
+          end
+          aggregation_state[key][:value] += value
+        end
+      end
+
+      def aggregate_timing(name, value, tags: [], no_prefix: false, type: DISTRIBUTION)
+        unless thread_healthcheck
+          sink << datagram_builder(no_prefix: no_prefix).d(name, value, CONST_SAMPLE_RATE, tags)
+          return
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, type)
+
+        mutex.synchronize do
+          if aggregation_state.key?(key) && aggregation_state[key][:value].size + 1 >= @max_values
+            do_flush
+          end
+          unless aggregation_state.key?(key)
+            aggregation_state[key] = {
+              type: type,
+              name: name,
+              value: [],
+              tags: tags,
+              no_prefix: no_prefix,
+            }
+          end
+          aggregation_state[key][:value] << value
+        end
+      end
+
+      def gauge(name, value, tags: [], no_prefix: false)
+        unless thread_healthcheck
+          sink << datagram_builder(no_prefix: no_prefix).g(name, value, CONST_SAMPLE_RATE, tags)
+          return
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, COUNT)
+
+        mutex.synchronize do
+          unless aggregation_state.key?(key)
+            aggregation_state[key] = {
+              type: GAUGE,
+              name: name,
+              value: value,
+              tags: tags,
+              no_prefix: no_prefix,
+            }
+          end
```

Comment:
```suggestion
          aggregation_state[key] ||= {
            type: GAUGE,
            name: name,
            value: value,
            tags: tags,
            no_prefix: no_prefix,
          }
        end
```

Reply:
Also it would be best to use a proper object rather than a Hash. Hashes aren't great for performance and compactness.

Reply:
@casperisfine would you recommend something like a `Struct` (or maybe the new `Data`, if we are to live on the edge)? Or is defining a regular class the way to go?

(I'd also love to know where I could read more about a `Hash` not being great for performance, I had always assumed it would be heavily optimized given how prevalent it is in ruby code...)

Reply:
A good old PORO.

> a Hash not being great for performance, I had always assumed it would be heavily optimized given how prevalent it is in ruby code...)

Hash is well optimized, but it's still a Hash lookup, whereas instance variables and method calls have inline caches that allow not to have to do a hash lookup in most cases. 

You can read more at https://railsatscale.com/2023-10-24-memoization-pattern-and-object-shapes/

Reply:
Using POROs now, please check if that is okay.